mindsdb_sdk/utils/openai.py

```diff
@@ -0,0 +1,257 @@
+    """
+    # You will need to pip install docstring-parser to use this function
+
+    import inspect
+    import docstring_parser
+
+    params = inspect.signature(function).parameters
+    docstring = docstring_parser.parse(function.__doc__)
+
+    # Get the first line of the docstring as the function description or use the user-provided description
+    function_description = description or docstring.short_description
+
+    function_dict = {
+        "type":"function",
+        "function":{
+            "name":function.__name__,
+            "description":function_description,
+            "parameters":{
+                "type":"object",
+                "properties":{},
+                "required":[]
+            }
+        }
+    }
+
+    for name, param in params.items():
+        param_description = next((p.description for p in docstring.params if p.arg_name == name), '')
+
+        # convert annotation type to string
+        if param.annotation is not inspect.Parameter.empty:
+            if inspect.isclass(param.annotation):
+                param_type = param.annotation.__name__
+            else:
+                param_type = str(param.annotation)
+        else:
+            param_type = None
+
+        function_dict["function"]["parameters"]["properties"][name] = {
+            "type":param_type,
+            "description":param_description
+        }
+
+        # Check if parameter is required
+        if param.default == inspect.Parameter.empty:
+            function_dict["function"]["parameters"]["required"].append(name)
+
+    return function_dict
+
+
+def make_mindsdb_tool(schema: dict) -> dict:
+    """
+    Make an OpenAI tool for querying a database connection in MindsDB
+
+    :param schema: database schema
+
+    :return: dictionary containing function metadata for openai tools
+    """
+    return {
+        "type":"function",
+        "function":{
+            "name":"query_database",
+            "description":"Use this function to answer user questions. Input should be a fully formed SQL query.",
+            "parameters":{
+                "type":"object",
+                "properties":{
+                    "query":{
+                        "type":"string",
+                        "description":f"""
+                                    SQL query extracting info to answer the user's question.
+                                    SQL should be written using this database schema:
+                                    {schema}
+                                    The query should be returned in plain text, not in JSON.
+                                    """,
+                    }
+                },
+                "required":["query"],
+            },
+        }
+    }
+
+
+def make_data_tool(
+    model: str,
+    data_source: str,
+    description: str,
+    connection_args: dict
+):
+    """
+    tool passing connection details for datasource to litellm callback
+
```

Comment:
This is the helper function that should work with litellm callback you created @tmichaeldb